src/main/kotlin/uk/gov/justice/digital/hmpps/prison/visitbooker/registry/service/BookerDetailsService.kt

```diff
@@ -0,0 +1,77 @@
+package uk.gov.justice.digital.hmpps.prison.visitbooker.registry.service
+
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.digital.hmpps.prison.visitbooker.registry.dto.BookerDto
+import uk.gov.justice.digital.hmpps.prison.visitbooker.registry.dto.CreateBookerDto
+import uk.gov.justice.digital.hmpps.prison.visitbooker.registry.exceptions.BookerNotFoundException
+import uk.gov.justice.digital.hmpps.prison.visitbooker.registry.model.entity.Booker
+import uk.gov.justice.digital.hmpps.prison.visitbooker.registry.model.entity.Prisoner
+import uk.gov.justice.digital.hmpps.prison.visitbooker.registry.model.entity.Visitor
+import uk.gov.justice.digital.hmpps.prison.visitbooker.registry.model.repository.BookerPrisonerRepository
+import uk.gov.justice.digital.hmpps.prison.visitbooker.registry.model.repository.BookerPrisonerVisitorRepository
+import uk.gov.justice.digital.hmpps.prison.visitbooker.registry.model.repository.BookerRepository
+
+@Service
+class BookerDetailsService(
+  private val bookerRepository: BookerRepository,
+  private val prisonerRepository: BookerPrisonerRepository,
+  private val visitorRepository: BookerPrisonerVisitorRepository,
+) {
+
+  @Transactional
+  fun getBooker(reference: String): Booker {
+    return bookerRepository.findByReference(reference) ?: throw BookerNotFoundException("Booker for reference : $reference not found")
+  }
+
+  @Transactional
+  fun createOrUpdateBookerDetails(createBookerDto: CreateBookerDto): BookerDto {
+    val booker = bookerRepository.findByEmail(createBookerDto.email)?.let {
+      // clear child objects from booker
+      it.prisoners.clear()
+      bookerRepository.saveAndFlush(it)
+    } ?: run {
+      bookerRepository.saveAndFlush(Booker(email = createBookerDto.email))
+    }
+
+    val saveBooker = createChildObjects(createBookerDto, booker)
+    return BookerDto(saveBooker)
+  }
+
+  @Transactional
+  fun clearBookerDetails(bookerReference: String): BookerDto {
+    val booker = getBooker(bookerReference)
+    booker.prisoners.clear()
+    return BookerDto(bookerRepository.saveAndFlush(booker))
+  }
+
+  private fun createChildObjects(
```

Comment:
Could these be in the PrisonerService & VisitorService?

Reply:
If anything PrisonerService and VisitorService should be all in this service as this is the entry and master service... I might do this in tech debt task..